[19.05] Fix input parameter comparisons

mvdbeek opened the pull request:
Fixes https://github.com/galaxyproject/tools-iuc/pull/2702#issuecomment-559718768
Broken in https://github.com/galaxyproject/galaxy/pull/7826

The implementation of `__gt__` and `__eq__` is simpler now. One possible breaking change is that a TextToolParameter with value "5" is not equal to a literal 5 anymore. I doubt anyone relied on this behavior.

mvdbeek committed:
Fix input parameter comparisons

Fixes https://github.com/galaxyproject/tools-iuc/pull/2702#issuecomment-559718768
Broken in https://github.com/galaxyproject/galaxy/pull/7826
Also compares via `float` instead of `int` when other is int,
since that is more reasonable. Otherwise 5.1 > 5 is False.

nsoranzo commented on lib/galaxy/tools/wrappers.py:
```diff
@@ -80,14 +80,12 @@ def __init__(self, input, value, other_values={}):
-        elif isinstance(other, int):
-            return int(self) == other
-        elif isinstance(other, float):
+        elif isinstance(other, (int, float)) or (isinstance(other, InputValueWrapper) and other.input.type in ('float', 'integer')):
```
I'd keep int and float separate, just to be sure.

mvdbeek replied:
I think we need to treat them the same, otherwise the last test I added (5 < 5.1) would fail, or did I miss something?

nsoranzo replied:
Wouldn't the following work?

```suggestion
        elif isinstance(other, int) or (isinstance(other, InputValueWrapper) and other.input.type == 'integer'):
            return int(self) == other
        elif isinstance(other, float) or (isinstance(other, InputValueWrapper) and other.input.type == 'float'):
```

mvdbeek replied:
If self wraps a float this would lead to wrong results, e.g 5.1 == 5. It would be ok if we made this 
```
elif self.input.type == 'integer' and  isinstance(other, int) or (isinstance(other, InputValueWrapper) and other.input.type == 'integer'):
```
But then we also need to take care of the case where `self.input.type == 'integer'` and other is float-ish ... .
On the other hand I think casting to float should work fine in most/all cases.

mvdbeek replied:
Or ... we turn this around and cast self to the correct type. That seems like less guesswork.

nsoranzo commented on lib/galaxy/tools/wrappers.py:
```diff
@@ -80,14 +80,12 @@ def __init__(self, input, value, other_values={}):
             return float(self) == other
         else:
-            return super(InputValueWrapper, self) == other
+            return self.value == other
```
Not sure I'd go that far.

mvdbeek replied:
We could instead raise an exception here, but this should work for comparing color or boolean parameters for instance.

nsoranzo replied:
I'd prefer to cross that bridge when there is a reported issue/use case, especially since this PR is targeting 19.05.

mvdbeek replied:
I agree with the sentiment, but what's your suggestion ? This doesn't work at all right now (none of the parent classes define `__gt__` or `__eq__`). I'd be in favor of casting self to string.

nsoranzo commented on lib/galaxy/tools/wrappers.py:
```diff
@@ -110,14 +108,12 @@ def __getattr__(self, key):
             return float(self) > other
         else:
-            super(InputValueWrapper, self).__gt__(other)
+            return self.value > other
```
Here too.

mvdbeek committed:
Simplify parameter value comparisons

Cast self instead of other. This fails 5 == '5', which is less magic
(https://github.com/galaxyproject/galaxy/pull/9047#discussion_r352111557)

nsoranzo commented on test/unit/tools/test_wrappers.py:
```diff
@@ -83,17 +85,40 @@ def test_raw_object_wrapper():
+    wrapper = valuewrapper(tool, 5, "integer")
+    assert str(wrapper) == valuewrapper(tool, "5", "text")
+    assert int(wrapper) == valuewrapper(tool, "5", "integer")
+    assert wrapper == valuewrapper(tool, "5", "text")
```
I think this should not be allowed, too much magic, even in normal Python `"5" == 5` is False.

mvdbeek replied:
But that has previously worked, see the test above that I didn't modify.

mvdbeek replied:
that said this naturally falls out of casting InputValueWrapper to InputValueWrapper.input.type (https://github.com/galaxyproject/galaxy/pull/9047/commits/155660083d66d00423c2ed5ef6ee9028badd4883), which maybe means we should do this ?

nsoranzo commented on test/unit/tools/test_wrappers.py:
```diff
@@ -83,17 +85,40 @@ def test_raw_object_wrapper():
     assert wrapper == 5.0
     assert wrapper > 2
     assert wrapper < 10
+    assert wrapper <= 5.1
```
```suggestion
    assert wrapper < 5.1
```

nsoranzo commented on test/unit/tools/test_wrappers.py:
```diff
@@ -83,17 +85,40 @@ def test_raw_object_wrapper():
+    assert wrapper == valuewrapper(tool, "5", "float")
+    assert wrapper > valuewrapper(tool, "2", "integer")
+    assert wrapper < valuewrapper(tool, "10", "integer")
+    assert wrapper <= valuewrapper(tool, "5.1", "float")
```
```suggestion
    assert wrapper < valuewrapper(tool, "5.1", "float")
```

nsoranzo commented on lib/galaxy/tools/wrappers.py:
```diff
@@ -80,14 +80,16 @@ def __init__(self, input, value, other_values={}):
-            return super(InputValueWrapper, self) == other
+        return self.get_cast_value() == other
+
+    def get_cast_value(self):
```
```suggestion
    def _get_cast_value(self):
```
unless you intended this to be exposed.

Also I'd move it before `__eq__()`.

nsoranzo reviewed (commented):
I like this much better, thanks!

mvdbeek committed:
Cleanup test_input_value_wrapper assertions

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Rename get_cast_value to _get_cast_value

mvdbeek commented:
> I like this much better, thanks!

Yep, thanks for the help and discussions.
